pkg/crds/crds.go

```diff
@@ -0,0 +1,205 @@
+		// only return CRDs that are specified and have been migrated to the new CRD generation flow
+		if MigratedResources[crdName] {
+			toReturn[crdName] = true
+		}
+	}
+
+	allCRDs, err := crdsFromDir(yamlDir)
+	if err != nil {
+		return nil, err
+	}
+
+	capiMap := toMap(CAPICRDs())
+	bootstrapFleetMap := toMap(bootstrapFleet())
+
+	retCRDs := []*apiextv1.CustomResourceDefinition{}
+
+	for _, crd := range allCRDs {
+		if !toReturn[crd.Name] {
+			// skip crds that we were not asked for or have already been found.
+			continue
+		}
+
+		// set the entry to false so we know we it was found.
+		toReturn[crd.Name] = false
+
+		if crd.Labels == nil {
+			crd.Labels = map[string]string{}
+		}
+
+		// dynamically add labels to capi crds since we do not generate them.
+		if capiMap[crd.Name] {
+			crd.Labels["auth.cattle.io/cluster-indexed"] = "true"
+		}
+
+		// Add managed by label
+		if bootstrapFleetMap[crd.Name] {
+			// Ensure labels/annotations are set so that helm will manage this
+			crd.Labels[k8sManagedByKey] = "Helm"
+			if crd.Annotations == nil {
+				crd.Annotations = map[string]string{}
+			}
+			crd.Annotations["meta.helm.sh/release-name"] = fleet.CRDChartName
+			crd.Annotations["meta.helm.sh/release-namespace"] = fleet.ReleaseNamespace
+		} else {
+			// since this CRD is installed by rancher add the manged by label
+			crd.Labels[k8sManagedByKey] = managerValue
+		}
+
+		retCRDs = append(retCRDs, crd)
+	}
+
+	return retCRDs, hasMissingCRDs(toReturn)
+}
+
+// crdsFromDir recursively traverses the embedded yaml directory and find all CRD yamls.
+func crdsFromDir(dirName string) ([]*apiextv1.CustomResourceDefinition, error) {
+	// read all entries in the embedded directory
+	crdFiles, err := crdFS.ReadDir(dirName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read embedded dir '%s': %w", yamlDir, err)
+	}
+
+	allCRDs := []*apiextv1.CustomResourceDefinition{}
+	for _, dirEntry := range crdFiles {
+		fullPath := filepath.Join(dirName, dirEntry.Name())
+		if dirEntry.IsDir() {
+			// if the entry is the dir recurse into that folder to get all crds
+			subCRDs, err := crdsFromDir(fullPath)
+			if err != nil {
+				return nil, err
+			}
+			allCRDs = append(allCRDs, subCRDs...)
+			continue
+		}
+		ext := strings.ToLower(filepath.Ext(fullPath))
+		if ext != ".yaml" && ext != ".yml" {
+			// skip non yaml files
+			continue
+		}
+
+		// read the file and convert it to a crd object
+		file, err := crdFS.Open(fullPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open embedded file '%s': %w", fullPath, err)
+		}
+		crdObjs, err := yaml.UnmarshalWithJSONDecoder[*apiextv1.CustomResourceDefinition](file)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert embedded file '%s' to yaml: %w", fullPath, err)
+		}
+		for _, crdObj := range crdObjs {
+			if crdObj.Kind != "CustomResourceDefinition" {
+				// If the yaml was not a CRD skip it.
+				continue
+			}
+			allCRDs = append(allCRDs, crdObj)
```

Comment:
Nit: You might consider only appending CRDs if they are one of the ones that we requested (and passing in the toReturn map to indicate what was requested). This would be easier for the consuming function to process and avoid loading in all of the CRD definitions in memory (particularly the ones that we aren't using, which could be KB-MB in size).

Reply:
> This still doesn't mean they will be removed from memory any sooner since it depends entirely on when the Garbage collector runs.

I think that this may be slightly incorrect, since in the proposed model we mark it for removal "sooner", so it's possible that an "earlier sweep" may catch this. But I think the difference here is quite negligible, so we can resolve this thread.

Reply:
I agree, I was trying to convey that it was not guaranteed sooner.